### DISCUSSION TITLE
Hardware Barrier client

### DISCUSSION THREAD

**Discussion by windoze** [debauchee/barrier#1924] [Show and tell]
Hi, All,

I just made a Barrier client running on ESP32S3, it turns Barrier into a hardware KVM which can support virtually any OS.

To whom it may interest, the project is at: https://github.com/windoze/esparrier-idf

It's developed and tested on [M5Atom S3 Lite](https://docs.m5stack.com/en/core/AtomS3%20Lite) but any ESP32S3 board with native USB port should work.

Although some features are still missing but the basic functions work smoothly. This very post is written with it.

Feel free to share your ideas and comments.

Regards,
Chen

**Comment by shymega** [comment]
This is exciting to see. You may want to look at Input Leap - the `rust` branch is where I'm experimenting with replacing C++ with Rust.

**Reply by windoze** [reply]
Thanks for your info.

I briefly checked the source code and found that InputLeap doesn't change communication protocols, so I think my program should work with InputLeap as well.
In my code I rewrote the protocol part in Rust but I didn't implement every detail such as drag-and-drop and grab-clipboard, also I'm not quite sure about the meaning of every option so I ignored most of them.
But the [protocol part](https://github.com/windoze/esparrier-idf/tree/main/src/barrier) should be reusable and extended in the new Rust version of Barrier/InputLeap, you may want to take a look and provide your comments.

**Comment by davidkennedydev** [comment]
Is it still using the network for the mouse connection?

**Reply by windoze** [reply]
Yes, of course. Isn't that the purpose of using Barrier? 😄 
The difference is, this hardware client uses its own WiFi to connect to the Barrier server so nothing needs to be installed on the computer, it acts exactly the same as real ordinary keyboard/mouse from the computer perspective.
If you have a persistent power supply, e.g. an external-powered USB hub or a ESP32S3 board with battery, it can work in BIOS/EFI as well.
